craft_parts/executor/part_handler.py

```diff
@@ -655,6 +733,35 @@ def _remove(filename: Path) -> None:
         shutil.rmtree(filename)
 
 
+def _apply_file_filter(
+    *, filter_files: Set[str], filter_dirs: Set[str], destdir: Path
+) -> None:
+    for (root, directories, files) in os.walk(destdir, topdown=True):
+        for file_name in files:
+            path = Path(root, file_name)
+            relpath = path.relative_to(destdir)
+            if str(relpath) not in filter_files and not overlays.is_whiteout_file(path):
+                logger.debug("delete file: %s", relpath)
+                path.unlink()
+
+        for directory in directories:
+            path = Path(root, directory)
+            relpath = path.relative_to(destdir)
+            if path.is_symlink():
+                if str(relpath) not in filter_files:
```

Comment:
If it's a symlink to a directory, the name shouldn't come in `filter_dirs`, not `filter_files`?

Reply:
It's added to `filter_files` because we want to remove the symlink itself (so it will be removed with `unlink()`), not the directory it points to.